### PR TITLE
Feature/improved turn logic

### PR DESCRIPTION
**Pull request by soccer-fan**
Reworked large chunks of TacticalActionStrategy, ActionOption, World and UnitInstance (from most to fewest changes left to right) to add much richer turn logic.

Side effects of this include implementing movement and changing how ActionOptions work in a big way. I also introduced copy constructs for World and Unit instances to allow ActionOptions to be applied in a sandbox.

All unit tests pass - there are more things I would like to do to further tidy and add fuller features but I want to get this PR in as it already has a large changelist as is.

### CODE REVIEW

File: src/main/kotlin/io/hsar/wh40k/combatsimulator/logic/actionoptions/AttackActionOption.kt [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+import io.hsar.wh40k.combatsimulator.model.unit.Attribute
+import io.hsar.wh40k.combatsimulator.model.unit.BodyPart
+import io.hsar.wh40k.combatsimulator.model.unit.NumericValue
+import io.hsar.wh40k.combatsimulator.random.AverageDice
+import io.hsar.wh40k.combatsimulator.random.RandomDice
+import io.hsar.wh40k.combatsimulator.random.RollResult
+
+abstract class AttackActionOption: ActionOption() {
+    /*
+    Provide some concrete implementations of common damage calculation tasks that can be re-used by child classes
```
Comment (HSAR) [inline src/main/kotlin/io/hsar/wh40k/combatsimulator/logic/actionoptions/AttackActionOption.kt:13]: This probably ought to be above the class declaration, not under it.

File: src/main/kotlin/io/hsar/wh40k/combatsimulator/model/World.kt [hunk 2]
```diff
@@ -29,13 +22,74 @@ data class World(
...
-                        .let { bonus ->
-                            moveType.getMovementRange(bonus)
-                        })
-                && moveType.isValidMovementPath(unitPositions.getValue(unit), unitPositions.getValue(otherUnit))
+    /*
+    Given a user and another UnitInstance that the user wants to move towards, this will handle the low level
+    mechanics of moving the unit as close as it can get for its movement range, stoppign if it comes within melee
+    distance
+     */
+    fun moveTowards(unit: UnitInstance, otherUnit: UnitInstance, maxMovement: Int): Unit {
```
Comment (HSAR) [inline src/main/kotlin/io/hsar/wh40k/combatsimulator/model/World.kt:30]: This implementation is... fairly fugly. I think I'll leave it be for now but this is a candidate for some serious reworking in the near future.

File: src/test/kotlin/TestUtils.kt [hunk 3]
```diff
@@ -43,4 +42,106 @@ object TestUtils {
                 unitPositions
         )
     }
+
+    fun getMockedBasicActions(): Map<String, ActionOption> {
+        val halfAimMock = Mockito.mock(HalfAim::class.java)
+        Mockito.`when`(halfAimMock.actionCost).thenReturn(ActionCost.HALF_ACTION)
```
Comment (HSAR) [inline src/test/kotlin/TestUtils.kt:48]: [Minor] Doing a static import of Mockito.when will cut a lot of this noise.

File: src/main/kotlin/io/hsar/wh40k/combatsimulator/random/AverageDice.kt [hunk 4]
```diff
@@ -9,4 +9,8 @@ object AverageDice {
     private val roller = DiceRoller(AverageNumberGenerator())
 
     fun roll(diceString: String): Int = parser.parse(diceString, roller).totalRoll
+
+    fun averageDegrees(skillTarget: Int) {
+
```
Comment (HSAR) [inline src/main/kotlin/io/hsar/wh40k/combatsimulator/random/AverageDice.kt:14]: Is this... correct?

File: src/main/kotlin/io/hsar/wh40k/combatsimulator/logic/TacticalActionStrategy.kt [hunk 5]
```diff
@@ -1,67 +1,124 @@
...
+                                        tempWorld.replaceUnitInstanceWithCopy(targetedAction.target)
+                                    } else {
+                                        tempUser  // already temp as we have copied the user
+                                    }
+                                            // TODO not a massive issue but if both actions of a combo have the same target then any changes to the target won't carry over
+                                            .let { tempTarget ->
+                                                if (targetedAction.action.isLegal(tempWorld, tempUser, tempTarget)) {
+                                                    targetedAction.action.expectedValue(tempWorld, tempUser, tempTarget)
+                                                            .also {
+                                                                targetedAction.action.apply(tempWorld, tempUser, tempTarget)
```
Comment (HSAR) [inline src/main/kotlin/io/hsar/wh40k/combatsimulator/logic/TacticalActionStrategy.kt:104]: Just thinking, if `apply()` returns the new World state maybe we could get simplify (even if we can't get totally around) the copy problem.

File: src/test/kotlin/io/hsar/wh40k/combatsimulator/content/ItemDatabaseTest.kt [hunk 6]
```diff
@@ -41,6 +36,33 @@ class ItemDatabaseTest {
...
+        assertThat(result.itemName, equalTo(expectedItem.itemName))
+        assertThat(result.itemType, equalTo(expectedItem.itemType))
+        assertThat(result.modifiesAttributes[WEAPON_TYPE], equalTo(expectedItem.modifiesAttributes[WEAPON_TYPE]))
+
+        // Check both expected and actual have one action available
+        assertThat(
+                (result.modifiesAttributes[ACTIONS] as ActionValue).value.size,
+                equalTo((expectedItem.modifiesAttributes[ACTIONS] as ActionValue).value.size))
+
+        // and that both expected and actual have the action as the same class of ActonOption
```
Comment (HSAR) [inline src/test/kotlin/io/hsar/wh40k/combatsimulator/content/ItemDatabaseTest.kt:49]: [Minor] Typo "ActonOption" should be "ActionOption"

File: src/test/kotlin/io/hsar/wh40k/combatsimulator/content/ItemDatabaseTest.kt [hunk 7]
```diff
@@ -41,6 +36,33 @@ class ItemDatabaseTest {
...
+        assertThat(
+                (result.modifiesAttributes[ACTIONS] as ActionValue).value[0]::class,
```
Comment (HSAR) [inline src/test/kotlin/io/hsar/wh40k/combatsimulator/content/ItemDatabaseTest.kt:51]: You probably want to use [this matcher](http://hamcrest.org/JavaHamcrest/javadoc/1.3/org/hamcrest/core/IsInstanceOf.html)

**Review by HSAR — DISMISSED** [review]
Broadly LGTM considering the huge number of LoC.

File: src/main/kotlin/io/hsar/wh40k/combatsimulator/CombatSimulation.kt [hunk 8]
```diff
@@ -9,6 +9,7 @@ import io.hsar.wh40k.combatsimulator.model.unit.BaseStat.AGILITY
 import io.hsar.wh40k.combatsimulator.model.unit.NumericValue
 import io.hsar.wh40k.combatsimulator.model.unit.StatUtils.getBonus
 import io.hsar.wh40k.combatsimulator.random.RandomDice.rollInitiative
+import kotlin.contracts.contract
```
[the remark this answers is not in the document]
Comment (markyMitch) [reply src/main/kotlin/io/hsar/wh40k/combatsimulator/CombatSimulation.kt:12]: probably not, I need to go through each file and wipe the unused imports. thanks for the reminder

[the remark this answers is not in the document]
Comment (markyMitch) [reply src/main/kotlin/io/hsar/wh40k/combatsimulator/CombatSimulation.kt]: All unused imports removed across the codebase

File: src/main/kotlin/io/hsar/wh40k/combatsimulator/model/World.kt (hunk 2 above)
Comment (markyMitch) [reply src/main/kotlin/io/hsar/wh40k/combatsimulator/model/World.kt:30]: I agree it's not functional and probably not the most efficient either. Should be an easy refactor in place

File: src/test/kotlin/TestUtils.kt (hunk 3 above)
Comment (markyMitch) [reply src/test/kotlin/TestUtils.kt:48]: Fixed in TestUtils

File: src/main/kotlin/io/hsar/wh40k/combatsimulator/random/AverageDice.kt (hunk 4 above)
Comment (markyMitch) [reply src/main/kotlin/io/hsar/wh40k/combatsimulator/random/AverageDice.kt]: Its an unused method that I must've forgotten to remove. It is now removed.

File: src/main/kotlin/io/hsar/wh40k/combatsimulator/logic/TacticalActionStrategy.kt (hunk 5 above)
Comment (markyMitch) [reply src/main/kotlin/io/hsar/wh40k/combatsimulator/logic/TacticalActionStrategy.kt:104]: That sounds like a good idea in principle, although you'd have to deep copy world to return it or you'd have the same issue due to the obj references in world pointing to the same things.

Comment (markyMitch) [reply src/main/kotlin/io/hsar/wh40k/combatsimulator/logic/TacticalActionStrategy.kt:104]: Perhaps could refactor in a future feature branch?

File: src/test/kotlin/io/hsar/wh40k/combatsimulator/content/ItemDatabaseTest.kt (hunk 6 above)
Comment (markyMitch) [reply src/test/kotlin/io/hsar/wh40k/combatsimulator/content/ItemDatabaseTest.kt]: Updated

File: src/test/kotlin/io/hsar/wh40k/combatsimulator/content/ItemDatabaseTest.kt (hunk 7 above)
Comment (markyMitch) [reply src/test/kotlin/io/hsar/wh40k/combatsimulator/content/ItemDatabaseTest.kt]: Updated to use that matcher.

File: src/main/kotlin/io/hsar/wh40k/combatsimulator/logic/actionoptions/AttackActionOption.kt (hunk 1 above)
Comment (markyMitch) [reply src/main/kotlin/io/hsar/wh40k/combatsimulator/logic/actionoptions/AttackActionOption.kt:12]: Updated

File: src/test/kotlin/TestUtils.kt [hunk 9]
```diff
@@ -1,18 +1,19 @@
...
 import io.hsar.wh40k.combatsimulator.model.World
-import io.hsar.wh40k.combatsimulator.model.unit.Attribute
-import io.hsar.wh40k.combatsimulator.model.unit.BaseStat
-import io.hsar.wh40k.combatsimulator.model.unit.DerivedStats
-import io.hsar.wh40k.combatsimulator.model.unit.NumericValue
-import io.hsar.wh40k.combatsimulator.model.unit.Stats
+import io.hsar.wh40k.combatsimulator.model.unit.*
 import io.hsar.wh40k.combatsimulator.model.unit.Unit
+import org.mockito.Mockito
+import org.mockito.Mockito.*
```
Comment (HSAR) [inline src/test/kotlin/TestUtils.kt:10]: [Minor] Generally wanting to avoid star imports but this is a test so IDGAF really

**Review by HSAR — DISMISSED** [review]
LGTM

**Review by HSAR — APPROVED** [review]
Be good for you to take a look at the diff as best you can. Ammo tracking at the moment is very much WIP (if you have two guns they somehow share magazines), and several effects aren't implemented (see PRIMITIVE).

**Pull request merged**